docs/logging-design.md

```diff
@@ -0,0 +1,51 @@
+</details>
+<details>
+  <summary>Each request will get a globally unique request ID (`ahc-request-id`) that will be attached (as metadata) to each log message relevant to a request.</summary>
+  When many concurrent requests are active, it can be challenging to figure out which log message is associated with which request. To facilitate this task, AsyncHTTPClient will add a metadata field `ahc-request-id` to each log message so you can first find the request ID that is causing issues and then filter only messages with that ID.
+</details>
+<details>
+  <summary>Your `Logger` metadata is preserved.</summary>
+  AsyncHTTPClient accepts a `Logger` on every request method. This means that all the metadata you have attached, will be present on log messages issued by AsyncHTTPClient.
+
+ For example, if you attach `["my-system-req-uuid": "84B453E0-0DFD-4B4B-BF22-3434812C9015"]` and then do two requests using AsyncHTTPClient, both of those requests will carry `"my-system-req-uuid` as well as AsyncHTTPClient's `ahc-request-id`. This allows you to filter all HTTP request made from one of your system's requests whilst still disambiguating the HTTP requests (they will have different `ahc-request-id`s.
+</details>
+<details>
+  <summary>Instead of accepting one `Logger` instance per `HTTPClient` instance, each request method can accept a `Logger`.</summary>
+  This allows AsyncHTTPClient to preserve your metadata and add its own metadata such as `ahc-request-id`.
+</details>
+<details>
+  <summary>Only static log messages are used.</summary>
+  None of the log messages issued by AsyncHTTPClient will use String interpolation which means they will always be the exact same message.
+
+  For example when AsyncHTTPClient wants to tell you it got an actual network connection to perform a request on, it will give the logger the following pieces of information:
+
+  - message: `got connection for request`
+  - metadata (the values are example):
+    - `ahc-request-id`: `0`
+    - `ahc-connection`: `SocketChannel { BaseSocket { fd=15 }, active = true, localAddress = Optional([IPv4]127.0.0.1/127.0.0.1:54459), remoteAddress = Optional([IPv4]127.0.0.1/127.0.0.1:54457) }`
+
+  As you can see above, the log message doesn't actually contain the request or the network connection. Both of those pieces of information are in the `metadata`.
+
+  The rationale is that many people use log aggregation systems where it may be very useful to aggregate by log message which is much simplified by using a constant string instead of one with the information (such as the connection) in the middle of the log message.
```

Comment:
It may also worth mentioning log filtering, from comments above it looks like more common use case.